src/plugins/dataflow/utilities/playback-utils.ts

```diff
@@ -0,0 +1,62 @@
+import { Node } from "rete";
+import { IDataSet } from "../../../models/data/data-set";
+import { kMaxNodeValues } from "../model/utilities/node";
+import { ValueControl } from "../nodes/controls/value-control";
+import { ICaseCreation } from "../../../models/data/data-set-types";
+import { InputValueControl } from "../nodes/controls/input-value-control";
+import { DemoOutputControl } from "../nodes/controls/demo-output-control";
+
+function getPriorCases(dataSet: IDataSet, playhead: number){
+  const offset = 1;
+  const maxValues = kMaxNodeValues + offset;
+  const pastCalc = playhead - maxValues;
+  const regionStart = pastCalc < 0 ? 0 : pastCalc;
+  const countOfCasesToGet = playhead < maxValues ? playhead : maxValues;
+  return dataSet.getCasesAtIndices(regionStart, countOfCasesToGet);
+}
+
+const asStrings = ["Generator", "Transform", "Math", "Control", "Timer", "Logic"];
+const asNumbers = ["Number", "Sensor"];
+const asOutputs = ["Demo Output", "Live Output"];
+
+export function runNodePlaybackUpdates(node: Node, valForNode: number){
```

Comment:
I think I did a bad job explaining what I was hoping would be fixed about this in my last review.

A big `switch` statement like you had before isn't ideal, but it's not the worst, especially given that the core dataflow infrastructure is not the best. I agree that sticking with the big switch case is probably our best option for now.*

The issue that I wanted to address was duplicating each node's output string in two places in code. For example, the normal output is specified for the control node on lines 97-101 of `control-rete-node-factory.ts`, but then is duplicated in `runNodePlaybackUpdates`. There should just be one function to translate the value into an output string for each node type, and that function should be called wherever we need that string.

Looking more closely at `runNodePlaybackUpdates`, it doesn't even look like the outputs match the normal outputs for the nodes. Is that intentional? If so, why? I feel like we want the playback to mimic the normal behavior as much as possible. If they should be the same in both cases, this is a perfect example of why we should be using a single function in both cases. That will make it possible to have identical output without having to carefully QA each possibility every time.

*Ideally what you'd want to do here is have each node have its own `setValueAndOutput` function which handles the value in whichever way makes sense for that node type. In that case, all `runNodePlaybackUpdates` would do would be call `node.setValueAndOutput(valForNode)`. However, I don't think a solution like that is possible given how poor the dataflow infrastructure is.